# Day 4 Reading Journal

This journal includes several required exercises, but it is meant to encourage active reading more generally.  You should use the journal to take detailed notes, catalog questions, and explore the content from Think Python deeply.

Reading: Think Python Chapter 10

**Due: Monday, January 30 at 12 noon**



## [Chapter 10](http://www.greenteapress.com/thinkpython2/html/thinkpython2011.html)

You may want to review [state diagrams in Chapter 2](http://greenteapress.com/thinkpython2/html/thinkpython2003.html). [Python Tutor](http://pythontutor.com/) is also helpful for visualizing the state of your program.



### Exercise 10.3 
Write a function called `middle` that takes a list and returns a new list that contains all but the first and last elements. So `middle([1,2,3,4])` should return `[2,3]`.

In [9]:
def middle(items):
    """
    Removes first and last item in a list
    Returns modified list
    >>> middle([1,2,3,4,5])
    [2, 3, 4]
    """
    del items[0]
    del items[len(items)-1]
    return items


import doctest
doctest.testmod(verbose=True)

Trying:
    middle([1,2,3,4,5])
Expecting:
    [2, 3, 4]
ok
1 items had no tests:
    __main__
1 items passed all tests:
   1 tests in __main__.middle
1 tests in 2 items.
1 passed and 0 failed.
Test passed.


TestResults(failed=0, attempted=1)

### Exercise 10.4 
Write a function called `chop` that takes a list, modifies it by removing the first and last elements, and returns `None`.

What is the difference between `middle` and `chop`? Sketch out the program state or take a look at each in Python Tutor and answer the question in the Markdown cell below.

In [28]:
def chop(items):
    """
    Removes first and last item in a list
    Returns None
    
    >>> chop([1,2,3,4,5])
    
    """
    del items[0]
    del items[len(items)-1]
    return None

import doctest
doctest.testmod(verbose=True)

Trying:
    chop([1,2,3,4,5])
Expecting nothing
ok
Trying:
    is_anagram("test","sett")
Expecting:
    True
ok
1 items had no tests:
    __main__
2 items passed all tests:
   1 tests in __main__.chop
   1 tests in __main__.is_anagram
2 tests in 3 items.
2 passed and 0 failed.
Test passed.


TestResults(failed=0, attempted=2)

Chop returns None

### Exercise 10.6 
Two words are anagrams if you can rearrange the letters from one to spell the other. Write a function called `is_anagram` that takes two strings and returns `True` if they are anagrams.

In [27]:
def is_anagram(word1, word2):
    """
    >>> is_anagram("test","sett")
    True
    """
    word1.lower()
    word2.lower()
    letter1 = list(word1)
    letter2 = list(word2)
    letter1.sort()
    letter2.sort()
    if letter1 == letter2:
        return True
    return False

import doctest
doctest.testmod()

TestResults(failed=0, attempted=2)

### Exercise 10.8  
The (so-called) Birthday Paradox: <br /><br />
1\. Write a function called `has_duplicates` that takes a list and returns `True` if there is any element that appears more than once. It should not modify the original list.

2\. If there are 23 students in your class, what are the chances that two of you have the same birthday? Put your answer in the Markdown cell below. You can estimate this probability by generating random samples of 23 birthdays and checking for matches. Hint: you can generate random birthdays with the randint function from the [random module](https://docs.python.org/2/library/random.html).

You can read about this problem at http://en.wikipedia.org/wiki/Birthday_paradox, and you can download Allen's solution from http://greenteapress.com/thinkpython2/code/birthday.py.

In [53]:
import random

def has_duplicates(items):
    for i in range(len(items)):
        for j in range(i+1, len(items)):
            if items[i] == items[j]:
                return True
    return False

count = 0
duplicate = []
for i in range(10000):
    students = []
    for j in range(23):
        students.append(random.randint(1,365))
    duplicate.append(has_duplicates(students))

for i in range(len(duplicate)):
    if duplicate[i]:
        count+=1
print("Number of classes with same birthday: ", count)

Number of classes with same birthday:  5040


About 50%, given that out of 10,000 trials, 5040 of them had duplicates.

### Challenge: Exercise 10.10 (optional)

You should read [Chapter 9.1](http://www.greenteapress.com/thinkpython2/html/thinkpython2010.html) and do Exercise 1 first.

To check whether a word is in the word list, you could use the `in` operator, but it would be relatively slow because it searches through the words in order (try it).

Because the words are in alphabetical order, we can speed things up with a bisection search (also known as binary search), which is similar to what you do when you look a word up in the dictionary. You start in the middle and check to see whether the word you are looking for comes before the word in the middle of the list. If so, then you search the first half of the list the same way. Otherwise you search the second half.

Either way, you cut the remaining search space in half. If the word list has 113,809 words, it will take about 17 steps to find the word or conclude that it’s not there.

Write a function called `bisect` that takes a sorted list and a target value and returns the index of the value in the list, if it’s there, or `None` if it’s not.

Or you could read the documentation of the `bisect` module and use that! Solution: http://greenteapress.com/thinkpython2/code/inlist.py.

## Reading Journal feedback

Have any comments on this Reading Journal? Feel free to leave them [here](https://goo.gl/forms/hZqCUAi4ir7hVN6x2) and we'll read them when you submit your journal entry. This could include suggestions to improve the exercises, topics you'd like to see covered in class next time, or other feedback.

If you have Python questions or run into problems while completing the reading, you should post them to Piazza instead so you can get a quick response before your journal is submitted.